sql-plugin/src/main/scala/com/nvidia/spark/rapids/spill/SpillFramework.scala

```diff
@@ -527,33 +622,74 @@ class SpillableColumnarBatchHandle private (
     materialized
   }
 
+
+  // used to gate when a spill is actively being done so that a second thread won't
+  // also begin spilling, and a handle won't release the underlying buffer if it's
+  // closed while spilling
+  private var toSpill: Option[ColumnarBatch] = None
+
+  // used to ensure the resource gets cleaned up properly after spilling and avoids
+  // double freeing
+  private var spilled: Option[ColumnarBatch] = None
+  override def releaseDeviceResource(): Unit = {
+    synchronized {
+      super.releaseDeviceResource()
+      spilled.foreach(_.close())
+      spilled = None
+    }
+  }
+
   override def spill(): Long = {
     if (!spillable) {
       0L
     } else {
-      synchronized {
-        if (host.isEmpty && dev.isDefined) {
-          withChunkedPacker { chunkedPacker =>
-            meta = Some(chunkedPacker.getPackedMeta)
-            host = Some(SpillableHostBufferHandle.createHostHandleWithPacker(chunkedPacker))
-          }
-          // We return the size we were created with. This is not the actual size
-          // of this batch when it is packed, and it is used by the calling code
-          // to figure out more or less how much did we free in the device.
-          approxSizeInBytes
+      val thisThreadSpills = synchronized {
+        if (host.isEmpty && dev.isDefined && toSpill.isEmpty) {
+          toSpill = dev
+          GpuColumnVector.incRefCounts(toSpill.get)
+          true
         } else {
-          0L
+          false
         }
       }
+      if (thisThreadSpills) {
+        withChunkedPacker { chunkedPacker =>
+          meta = Some(chunkedPacker.getPackedMeta)
+          var staging: Option[SpillableHostBufferHandle] =
+            Some(SpillableHostBufferHandle.createHostHandleWithPacker(chunkedPacker))
+          synchronized {
+            if (closed) {
+              staging.foreach(_.close())
+              staging = None
+              doClose()
+            } else {
+              host = staging
+            }
+            // set spilled to dev instead of toSpill so that if dev was already closed during spill,
```

Comment:
I am a little confused by this comment. `toSpill` _is_ `dev`. It's just a reference to it. So when you increased the refcount at line 649, you increased it for `dev` and `toSpill`. I don't believe it makes much of a difference what we set `spilled` to (dev or toSpill) but we should just remove the comment I think.

Reply:
The logic here is that it's an issue of timing: yes we set `toSpill` to equal `dev` but later if we call close, we intentionally leave `toSpill` as is in case we are mid spill, but set `dev` to `None`, therefore the two variables are no longer equal at that point.

Having said that, I think I may be able to simplify this now that we explicitly keep track of `closed` but let me check.

Reply:
Yeah I just meant the comment was confusing, understood that the reference `dev` becomes `None`. It's just more of, at the time of the comment, `dev` and `toSpill` are one and the same.